File: plugins/sysdeck-photos/photos.js
/*
 * SysDeck - Photos Panel (v0.0.35)
 * Author: Jeremy Anderson (https://dcos.net)
 *
 * v0.0.35 directive: "as well as a photo manager of equal quality.
 * with its own module." Following the Jellyfin pattern: each
 * supported backend ships as a systemd service with a built-in admin
 * web UI; the bridge starts/stops/restarts the service via systemctl
 * and the panel iframes the running admin UI. Equal quality means
 * the same service-control + iframe-load shape as Jellyfin.
 *
 * Multi-backend design — same shape as the DB Control module
 * (plugins/sysdeck-db/db.js): the operator chooses the backend
 * installed on the host; the bridge auto-detects each backend via
 * CLI availability and/or systemd unit presence.
 *
 * Supported backends (see bridge/photos.py BACKEND_REGISTRY):
 *   PhotoPrism         — single Go binary · MIT · port 2342
 *   Piwigo             — single PHP-FPM app · GPL-2.0 · port 80
 *   Lychee             — single PHP-FPM app · MIT · port 80
 *   Nextcloud Memories — Nextcloud plugin · AGPL-3.0 · port 80
 *   LibrePhotos        — Django + React · MIT · port 3000
 *
 * Excluded (intentionally): NFS-mounted photo libraries (no admin
 * panel — use the Remote FS module for storage management), Amanda
 * (backup system, not a photo manager), Google Photos / iCloud /
 * etc. (cloud-only, no on-host admin panel reachable).
 *
 * Bridge surface (see shared/bridge.js → bridge.photos):
 *   summary()              → {backends, totalBackends, installedCount, runningCount}
 *   status(id)             → single-backend detail dict
 *   start(id)              → {action, rc, success, output, stderr}
 *   stop(id)               → same shape
 *   restart(id)            → same shape
 *   webStatus(id)          → {running, url, port, ...}
 */

export async function mount(panel, { bridge, EventBus }) {
    panel.innerHTML = renderSkeleton();

    let summary = {};
    try {
        summary = await bridge.photos.summary();
    } catch (err) {
        panel.innerHTML = renderError(err);
        return;
    }

    const backends = summary.backends || [];
    const installed = backends.filter((b) => b.status !== 'uninstalled');
    const running = backends.filter((b) => b.status === 'running');

    panel.innerHTML = `
        <header>
            <h2 class="suite-panel-title">Photo Manager</h2>
            <p class="suite-panel-subtitle">
                Self-hosted photo & album management · multi-backend
                · ${summary.totalBackends || backends.length} supported
                · ${summary.installedCount || installed.length} installed
                · ${summary.runningCount || running.length} running
            </p>
        </header>

        <div class="suite-row">
            <div class="suite-card suite-col-3">
                <h3 class="suite-card-title">Backends</h3>
                <div class="suite-stat-value">${summary.totalBackends || backends.length}</div>
                <div class="suite-stat-label">total supported</div>
            </div>
            <div class="suite-card suite-col-3">
                <h3 class="suite-card-title">Installed</h3>
                <div class="suite-stat-value ${installed.length ? 'suite-warn' : ''}">${installed.length}</div>
                <div class="suite-stat-label">${installed.length ? 'detected on host' : 'none — install one'}</div>
            </div>
            <div class="suite-card suite-col-3">
                <h3 class="suite-card-title">Running</h3>
                <div class="suite-stat-value ${running.length ? 'suite-warn' : ''}">${running.length}</div>
                <div class="suite-stat-label">${running.length ? 'admin UIs live' : 'all stopped'}</div>
            </div>
        </div>

        ${renderBackendsGrid(backends, { bridge, EventBus })}

        <div class="suite-card">
            <div class="suite-card-body">
                <p class="suite-muted" style="font-size:0.85rem">
                    The bridge runs <code>systemctl start/stop/restart</code> for
                    the chosen backend via the cockpit superuser channel (polkit
                    <code>org.sysdeck.photos.modify</code>). The admin UI loads in
                    an iframe once the service is running. Each backend is a
                    separate process — no third-party code is bundled in the suite.
                </p>
            </div>
        </div>
    `;

    wireEvents(panel, { bridge, EventBus });
    EventBus.emit('photos.loaded', { installed: installed.length, running: running.length });
}

// ── Backend cards ──────────────────────────────────────────────────

function renderBackendsGrid(backends, { bridge, EventBus }) {
    if (!backends.length) {
        return `
            <div class="suite-card">
                <h3 class="suite-card-title">Backends (0)</h3>
                <div class="suite-card-body">
                    <p class="suite-muted">No photo backends registered.</p>
                </div>
            </div>
        `;
    }
    return backends.map((b) => renderBackendCard(b)).join('');
}

function renderBackendCard(b) {
    if (b.status === 'uninstalled') {
        return `
            <div class="suite-card">
                <div class="suite-card-header">
                    <h3 class="suite-card-title">${escapeHtml(b.name)} <span class="suite-muted" style="font-size:0.85rem">· ${escapeHtml(b.id)}</span></h3>
                    <span class="suite-badge">${escapeHtml(b.family)}</span>
                </div>
                <div class="suite-card-body">
                    <p class="suite-muted" style="font-size:0.85rem">
                        Not installed. ${escapeHtml(b.name)} is <strong>${escapeHtml(b.license)}</strong>
                        licensed — <a href="${escapeHtml(b.homepage)}" target="_blank" style="color:var(--sysdeck-accent)">${escapeHtml(b.homepage)}</a>
                    </p>
                    <p class="suite-muted suite-mono" style="margin-top:0.5rem;font-size:0.8rem">${escapeHtml(b.installHint)}</p>
                </div>
            </div>
        `;
    }

    const running = b.status === 'running';
    return `
        <div class="suite-card">
            <div class="suite-card-header">
                <h3 class="suite-card-title">${escapeHtml(b.name)} <span class="suite-muted" style="font-size:0.85rem">· ${escapeHtml(b.id)}</span></h3>
                <div class="suite-row" style="gap:0.5rem">
                    <button class="suite-btn ${running ? '' : 'suite-btn-primary'} btn-photos-start" data-id="${escapeHtml(b.id)}" ${running ? 'disabled' : ''}>▶ Start</button>
                    <button class="suite-btn btn-photos-stop" data-id="${escapeHtml(b.id)}" ${!running ? 'disabled' : ''}>■ Stop</button>
                    <button class="suite-btn btn-photos-restart" data-id="${escapeHtml(b.id)}" ${!running ? 'disabled' : ''}>↻ Restart</button>
                    <button class="suite-btn suite-btn-ghost btn-photos-refresh" data-id="${escapeHtml(b.id)}">↻</button>
                </div>
            </div>
            <div class="suite-card-body">
                <table class="suite-table">
                    <tbody>
                        <tr><th>Status</th><td>${statusBadge(b.status)}</td></tr>
                        <tr><th>Service</th><td class="suite-table-mono">${escapeHtml(b.serviceUnit || '—')}</td></tr>
                        <tr><th>Port</th><td class="suite-table-mono">${b.port}</td></tr>
                        <tr><th>URL</th><td class="suite-table-mono">${running ? `<a href="${escapeHtml(b.url)}" target="_blank" style="color:var(--sysdeck-accent)">${escapeHtml(b.url)}</a>` : '— (service stopped)'}</td></tr>
                        <tr><th>License</th><td class="suite-muted">${escapeHtml(b.license)} · <a href="${escapeHtml(b.homepage)}" target="_blank" style="color:var(--sysdeck-accent)">${escapeHtml(b.homepage)}</a></td></tr>
                    </tbody>
                </table>
                ${running ? renderWebIframe(b) : `<p class="suite-muted" style="font-size:0.85rem;margin-top:0.5rem">Click <strong>▶ Start</strong> to launch ${escapeHtml(b.name)}. The admin UI will load in an iframe below.</p>`}
            </div>
        </div>
    `;
}

function renderWebIframe(b) {
    return `
        <div style="margin-top:1rem">
            <iframe src="${escapeHtml(b.url)}"
                    style="width:100%;height:600px;border:1px solid var(--sysdeck-border);border-radius:6px;background:#1e1e1e;"
                    id="photos-iframe-${escapeHtml(b.id)}"
                    title="${escapeHtml(b.name)} Admin Panel"
                    allow="autoplay; fullscreen; encrypted-media; picture-in-picture"
                    allowfullscreen></iframe>
        </div>
    `;
}

// ── Event wiring ────────────────────────────────────────────────────

function wireEvents(panel, { bridge, EventBus }) {
    const action = async (btn, method, label) => {
        const id = btn.dataset.id;
        if (!id) return;
        btn.disabled = true;
        const original = btn.textContent;
        btn.textContent = `${label} ...`;
        try {
            const r = await bridge.photos[method](id);
            if (r?.error) {
                alert(`${label} ${id} failed:\n${r.error}`);
            } else if (!r?.success && method !== 'restart') {
                alert(`${label} ${id} failed:\n${r?.stderr || r?.output || 'unknown'}`);
            }
            setTimeout(() => mount(panel, { bridge, EventBus }), 1000);
        } catch (err) {
            btn.disabled = false;
            btn.textContent = original;
            alert(`${label} ${id} error: ${err.message || err}`);
        }
    };

    panel.querySelectorAll('.btn-photos-start').forEach((btn) => {
        btn.addEventListener('click', (ev) => action(ev.currentTarget, 'start', 'Start'));
    });
    panel.querySelectorAll('.btn-photos-stop').forEach((btn) => {
        btn.addEventListener('click', (ev) => action(ev.currentTarget, 'stop', 'Stop'));
    });
    panel.querySelectorAll('.btn-photos-restart').forEach((btn) => {
        btn.addEventListener('click', (ev) => action(ev.currentTarget, 'restart', 'Restart'));
    });
    panel.querySelectorAll('.btn-photos-refresh').forEach((btn) => {
        btn.addEventListener('click', () => {
            mount(panel, { bridge, EventBus });
        });
    });
}

// ── Utilities ───────────────────────────────────────────────────────

function statusBadge(status) {
    const map = {
        'running':     '<span class="suite-badge success">running</span>',
        'starting':    '<span class="suite-badge warn">starting</span>',
        'stopped':     '<span class="suite-badge">stopped</span>',
        'error':       '<span class="suite-badge danger">error</span>',
        'uninstalled': '<span class="suite-badge">uninstalled</span>',
        'unknown':     '<span class="suite-badge">unknown</span>',
    };
    return map[status] || `<span class="suite-badge">${escapeHtml(status || 'unknown')}</span>`;
}

function escapeHtml(s) {
    return String(s == null ? '' : s)
        .replace(/&/g, '&amp;')
        .replace(/</g, '&lt;')
        .replace(/>/g, '&gt;')
        .replace(/"/g, '&quot;');
}

function renderSkeleton() {
    return `<div class="suite-skeleton">
        <div class="suite-skeleton-line w-1/3"></div>
        <div class="suite-skeleton-line w-2/3"></div>
        <div class="suite-skeleton-line w-1/2"></div>
    </div>`;
}

function renderError(err) {
    return `<div class="suite-card">
        <h3 class="suite-card-title">Photos bridge unavailable</h3>
        <p class="suite-card-body suite-muted">${err.message || err}. Ensure the bridge helper is installed at /usr/lib/sysdeck/bridge/photos.py.</p>
    </div>`;
}

File: bridge/photos.py
#!/usr/bin/env python3
"""SysDeck - Photos Bridge Helper
Author: Jeremy Anderson (https://dcos.net)

Manages self-hosted photo management backends as systemd services
and exposes the built-in admin web UI for iframe embedding in the
SysDeck panel.

v0.0.35 directive: "as well as a photo manager of equal quality.
with its own module." Following the Jellyfin pattern: start/stop/
restart the service via systemctl; the panel iframes the running
admin web UI. Equal quality means the photo manager module ships
with the same service-control + iframe-load shape as Jellyfin.

Multi-backend design — same shape as bridge/db.py: the operator
chooses the backend that's installed on the host. Each entry in
BACKEND_REGISTRY declares its systemd unit, web port, install hint,
and license. The bridge auto-detects which are installed; the
panel renders a backend card per installed one and an install
hint card per absent one.

Backends shipped:
  PhotoPrism         — single Go binary · MIT · port 2342
  Piwigo             — single PHP-FPM app · GPL-2.0 · port 80
  Lychee             — single PHP-FPM app · MIT · port 80
  Nextcloud-Memories — Nextcloud plugin · AGPL-3.0 · port 80
  LibrePhotos        — Django + React · MIT · port 3000

Excluded (intentionally, per the "equal quality" bar):
  Google Photos / iCloud / etc. — cloud-only, no systemd unit,
  no admin panel reachable from the host.

Subcommands:
  summary       — list all detected backends with status + port + url
  status <id>   — detailed status for one backend
  start <id>    — systemctl start <service>
  stop <id>     — systemctl stop <service>
  restart <id>  — systemctl restart <service>
  web-status <id> — {running, port, url} for iframe embedding

Cockpit way (v0.0.31+ pattern): the bridge runs systemctl via
subprocess directly — no `sudo` shell-out. The JS panel passes
{ superuser: 'try' } to cockpit.spawn so the cockpit bridge prompts
the operator via polkit for the org.sysdeck.photos.modify action
(added in v0.0.35).

Each backend is invoked as a separate process via subprocess —
the suite (MIT) and each backend (its own license) remain
independent programs. No backend code is bundled.

Usage:
    python3 /usr/lib/sysdeck/bridge/photos.py summary
    python3 /usr/lib/sysdeck/bridge/photos.py start photoprism
    python3 /usr/lib/sysdeck/bridge/photos.py web-status photoprism
"""

import json
import shutil
import subprocess
import sys
from datetime import datetime
from typing import Any


# ── Backend Registry ───────────────────────────────────────────────
# Each entry: (id, name, family, default_port, systemd_unit, cli_tool,
#              config_paths, web_path, license, homepage, install_hint)

BACKEND_REGISTRY = [
    (
        "photoprism", "PhotoPrism", "go-binary", 2342,
        "photoprism.service", "photoprism",
        ["/etc/photoprism/options.yml", "/var/lib/photoprism/"],
        "/", "MIT",
        "https://github.com/photoprism/photoprism",
        "Arch: yay -S photoprism  ·  Debian: docker run photoprism/photoprism  ·  Fedora: docker run photoprism/photoprism",
    ),
    (
        "piwigo", "Piwigo", "php-app", 80,
        "php-fpm.service", "piwigo",
        ["/etc/piwigo/", "/usr/share/webapps/piwigo/"],
        "/piwigo/", "GPL-2.0",
        "https://github.com/Piwigo/Piwigo",
        "Arch: yay -S piwigo  ·  Debian: install under /var/www/piwigo + apache2 + php-fpm  ·  Fedora: same",
    ),
    (
        "lychee", "Lychee", "php-app", 80,
        "php-fpm.service", "lychee",
        ["/etc/lychee/", "/usr/share/webapps/lychee/"],
        "/lychee/", "MIT",
        "https://github.com/LycheeOrg/Lychee",
        "Arch: yay -S lychee  ·  Debian: install under /var/www/lychee + apache2 + php-fpm",
    ),
    (
        "nextcloud-memories", "Nextcloud Memories", "nextcloud-plugin", 80,
        "php-fpm.service", "occ",
        ["/etc/webapps/nextcloud/", "/usr/share/webapps/nextcloud/"],
        "/nextcloud/index.php/apps/memories/", "AGPL-3.0",
        "https://github.com/pulsejet/memories",
        "Arch: pacman -S nextcloud + occ app:enable memories  ·  Debian: apt install nextcloud-server",
    ),
    (
        "librephotos", "LibrePhotos", "django-react", 3000,
        "librephotos.service", "librephotos",
        ["/etc/librephotos/", "/var/lib/librephotos/"],
        "/", "MIT",
        "https://github.com/LibrePhotos/librephotos",
        "Arch: yay -S librephotos  ·  Debian: docker run librephotos/librephotos  ·  Fedora: docker run librephotos/librephotos",
    ),
]


def _have(binary: str) -> bool:
    return shutil.which(binary) is not None


def _unit_loaded(unit: str) -> bool:
    if not unit:
        return False
    out = subprocess.run(
        ["systemctl", "list-unit-files", unit],
        capture_output=True, text=True, timeout=5,
    ).stdout
    return unit in out


def _systemctl_show(unit: str, props: list[str]) -> dict[str, str]:
    out = subprocess.run(
        ["systemctl", "show", unit, "--property=" + ",".join(props)],
        capture_output=True, text=True, timeout=5,
    ).stdout.strip()
    result = {}
    for line in out.splitlines():
        k, _, v = line.partition("=")
        if k:
            result[k] = v
    return result


def _service_status(unit: str) -> dict[str, Any]:
    if not unit:
        return {"status": "uninstalled", "active": "", "sub": "", "uptime_seconds": 0}
    props = _systemctl_show(unit, ["ActiveState", "SubState", "ActiveEnterTimestamp"])
    active = props.get("ActiveState", "unknown")
    sub = props.get("SubState", "unknown")
    uptime = 0
    ts = props.get("ActiveEnterTimestamp", "")
    if ts:
        try:
            dt = datetime.strptime(ts[:25], "%a %Y-%m-%d %H:%M:%S")
            uptime = int((datetime.now() - dt).total_seconds())
        except (ValueError, OSError):
            pass
    if active == "active":
        status = "running"
    elif active == "activating":
        status = "starting"
    elif active == "failed":
        status = "error"
    elif active in ("inactive", "deactivating"):
        status = "stopped"
    else:
        status = "unknown"
    return {"active": active, "sub": sub, "status": status, "uptime_seconds": uptime}


def detect_backend(entry) -> dict[str, Any]:
    (bid, name, family, port, unit, cli, configs, web_path,
     lic, homepage, install_hint) = entry

    cli_available = _have(cli) if cli else False
    unit_loaded = _unit_loaded(unit) if unit else False

    if unit_loaded:
        state = _service_status(unit)
        status = state["status"]
        active_state = state["active"]
        sub_state = state["sub"]
        uptime = state["uptime_seconds"]
    elif cli_available:
        status = "stopped"
        active_state = ""
        sub_state = ""
        uptime = 0
    elif any(_path_exists(p) for p in configs):
        # Config dir present but service not registered — best-effort.
        status = "stopped"
        active_state = ""
        sub_state = ""
        uptime = 0
    else:
        status = "uninstalled"
        active_state = ""
        sub_state = ""
        uptime = 0

    config_path = next((p for p in configs if _path_exists(p)), "")

    return {
        "id": bid,
        "name": name,
        "family": family,
        "status": status,
        "active": active_state,
        "sub": sub_state,
        "uptime_seconds": uptime,
        "port": port,
        "webPath": web_path,
        "url": f"http://127.0.0.1:{port}{web_path}",
        "serviceUnit": unit,
        "cli": cli,
        "configPath": config_path,
        "license": lic,
        "homepage": homepage,
        "installHint": install_hint,
        "supported": True,
    }


def _path_exists(p: str) -> bool:
    try:
        return bool(p) and __import__("os").path.exists(p)
    except (OSError, ValueError):
        return False


def cmd_summary() -> dict[str, Any]:
    backends = [detect_backend(e) for e in BACKEND_REGISTRY]
    installed = [b for b in backends if b["status"] != "uninstalled"]
    running = [b for b in backends if b["status"] == "running"]
    return {
        "backends": backends,
        "totalBackends": len(backends),
        "installedCount": len(installed),
        "runningCount": len(running),
    }


def _find_backend(bid: str):
    for e in BACKEND_REGISTRY:
        if e[0] == bid:
            return e
    return None


def cmd_status(bid: str) -> dict[str, Any]:
    e = _find_backend(bid)
    if not e:
        return {"error": f"Unknown backend: {bid}"}
    return detect_backend(e)


def _systemctl(action: str, unit: str) -> dict[str, Any]:
    if not unit:
        return {"rc": 127, "success": False, "stderr": f"no systemd unit for this backend"}
    try:
        r = subprocess.run(
            ["systemctl", action, unit],
            capture_output=True, text=True, timeout=30,
        )
        return {
            "action": action,
            "service": unit,
            "rc": r.returncode,
            "success": r.returncode == 0,
            "output": (r.stdout or "").strip(),
            "stderr": (r.stderr or "").strip(),
        }
    except (subprocess.TimeoutExpired, FileNotFoundError, OSError) as exc:
        return {"action": action, "service": unit, "rc": 1,
                "success": False, "stderr": str(exc)}


def cmd_start(bid: str) -> dict[str, Any]:
    e = _find_backend(bid)
    if not e:
        return {"error": f"Unknown backend: {bid}"}
    return _systemctl("start", e[4])


def cmd_stop(bid: str) -> dict[str, Any]:
    e = _find_backend(bid)
    if not e:
        return {"error": f"Unknown backend: {bid}"}
    return _systemctl("stop", e[4])


def cmd_restart(bid: str) -> dict[str, Any]:
    e = _find_backend(bid)
    if not e:
        return {"error": f"Unknown backend: {bid}"}
    return _systemctl("restart", e[4])


def cmd_web_status(bid: str) -> dict[str, Any]:
    e = _find_backend(bid)
    if not e:
        return {"error": f"Unknown backend: {bid}"}
    (bid2, name, family, port, unit, cli, configs, web_path,
     lic, homepage, install_hint) = e
    state = _service_status(unit) if unit else {"status": "uninstalled"}
    return {
        "id": bid2,
        "name": name,
        "running": state["status"] == "running",
        "status": state["status"],
        "port": port,
        "webPath": web_path,
        "url": f"http://127.0.0.1:{port}{web_path}",
        "license": lic,
        "homepage": homepage,
        "installHint": install_hint,
    }


def main(argv: list[str]) -> int:
    if not argv or argv[0] in ("-h", "--help"):
        print(__doc__)
        return 0

    cmd = argv[0]
    if cmd == "summary":
        print(json.dumps(cmd_summary(), indent=2))
    elif cmd == "status":
        bid = argv[1] if len(argv) > 1 else ""
        print(json.dumps(cmd_status(bid), indent=2))
    elif cmd == "start":
        bid = argv[1] if len(argv) > 1 else ""
        print(json.dumps(cmd_start(bid), indent=2))
    elif cmd == "stop":
        bid = argv[1] if len(argv) > 1 else ""
        print(json.dumps(cmd_stop(bid), indent=2))
    elif cmd == "restart":
        bid = argv[1] if len(argv) > 1 else ""
        print(json.dumps(cmd_restart(bid), indent=2))
    elif cmd == "web-status":
        bid = argv[1] if len(argv) > 1 else ""
        print(json.dumps(cmd_web_status(bid), indent=2))
    else:
        print(json.dumps({"error": f"Unknown command: {cmd}"}))
        return 2
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))
